projects/chaos/src/charger.c

```diff
@@ -0,0 +1,45 @@
+#include "charger.h"
+
+#include <stdlib.h>
+
+#include "can.h"
+#include "can_msg_defs.h"
+#include "can_transmit.h"
+#include "can_unpack.h"
+#include "status.h"
+
+typedef struct ChargerStorage {
+  EEChargerSetRelayState relay_state;
+  EEChargerConnState conn_state;
+} ChargerStorage;
+
+static ChargerStorage s_storage;
+
+// CanRxHandlerCb
+static StatusCode prv_handle_charger_conn_state(const CANMessage *msg, void *context,
+                                                CANAckStatus *ack_reply) {
+  (void)context;
+  (void)ack_reply;
+  CAN_UNPACK_CHARGER_CONN_STATE(msg, (uint8_t *)&s_storage.conn_state);
+  charger_set_state(s_storage.relay_state);
+  return STATUS_CODE_OK;
+}
+
+StatusCode charger_init(void) {
+  s_storage.relay_state = EE_CHARGER_SET_RELAY_STATE_OPEN;
+  s_storage.conn_state = EE_CHARGER_CONN_STATE_DISCONNECTED;
+
+  return can_register_rx_handler(SYSTEM_CAN_MESSAGE_CHARGER_CONN_STATE,
+                                 prv_handle_charger_conn_state, NULL);
+}
+
+StatusCode charger_set_state(EEChargerSetRelayState state) {
```

Comment:
So based on the notify module, you need to call this repeatedly?

Reply:
This will be called in the main event loop to initially set the state (charging vs not charging) via the sequencer and from then on will be invoked each time the charger notifies.